### PR TITLE
Periodically check if new InfluxDB DBRPs have been created #553

### PR DESCRIPTION
**Pull request by titilambert**
###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] Tests pass
- [ ] CHANGELOG.md updated
- [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [ ] [TICKscript Spec](https://github.com/influxdata/kapacitor/blob/master/tick/TICKscript.md) updated

### CODE REVIEW

File: services/influxdb/service.go [hunk 1]
```diff
@@ -187,7 +189,14 @@ type subInfo struct {
...
 	if !s.disableSubs {
-		return s.linkSubscriptions()
+		err := s.linkSubscriptions()
+		ticker := time.NewTicker(s.subscriptionsRetryInterval)
+		go func() {
+			for _ = range ticker.C {
+				s.linkSubscriptions()
+			}
+		}()
+		return err
```
Comment (nathanielc) [inline services/influxdb/service.go]: Can you add the matching logic to stop the ticker in the `Close` method?

**Comment by nathanielc** [comment]
@titilambert Overall looks good. A few comments

1.  I think we should rename the variable SubscriptionRetryInterval to SubscriptionSyncInterval since its not retrying a failed task but rather polling for new information. 
2. We should check if the interval is 0 thus disabling subscription syncing. 
3. The current implementation of `linkSubscriptions` assumes that no services are running. But that is no longer the case, we need to update `linkSubscriptions` to keep track of which subs are already running and which are not.

**Comment by nathanielc** [comment]
Here at this line https://github.com/titilambert/kapacitor/blob/c673874f60cfa5123bca73b046f5da419468fb03/services/influxdb/service.go#L417

We need to check if the UDP service is already running before starting. A map[subEntry]bool on the `influxdb` struct should work.

**Comment by titilambert** [comment]
@nathanielc I fixed comment 1 and 2
About comment 3: I'm not sure to understand what is purpose of `startedSubs` variable

**Comment by titilambert** [comment]
@nathanielc I made a new commit which should include all your comments

**Comment by nathanielc** [comment]
@titilambert Looking good, I think you also need to mark the sub as running [here](https://github.com/titilambert/kapacitor/blob/41926fbdd3324c1fed9359f7cc04e4fb22bea917/services/influxdb/service.go#L409) too. The difference between those two loops is whether the sub has been created or not. We could probably simplify the loops a bit, but for now just keeping track in both loops should work.

**Comment by titilambert** [comment]
@nathanielc Done !

File: services/influxdb/service.go [hunk 2]
```diff
@@ -144,22 +145,25 @@ func (s *Service) NewNamedClient(name string) (client.Client, error) {
...
+	httpPort                 int
+	logger                   *log.Logger
+	protocol                 string
+	udpBind                  string
+	udpBuffer                int
+	udpReadBuffer            int
+	startupTimeout           time.Duration
+	subscriptionSyncInterval time.Duration
+	disableSubs              bool
+	runningSubs              map[subEntry]bool
```
Comment (nathanielc) [inline services/influxdb/service.go]: This map needs to be initialized or you get a `panic: assignment to entry in nil map`.

File: services/influxdb/service.go [hunk 3]
```diff
@@ -187,13 +191,25 @@ type subInfo struct {
...
+			}()
+		}
+		return err
 	}
 	return nil
 }
 
 func (s *influxdb) Close() error {
 	var lastErr error
+	if s.subscriptionSyncInterval != time.Duration(0) {
```
Comment (nathanielc) [inline services/influxdb/service.go]: I prefer `if s.subscriptionSyncInterval != 0` as there is less to read and as a result reads cleaner.

File: services/influxdb/service.go [hunk 4]
```diff
@@ -187,13 +191,25 @@ type subInfo struct {
 
 func (s *influxdb) Open() error {
 	if !s.disableSubs {
-		return s.linkSubscriptions()
+		err := s.linkSubscriptions()
+		if s.subscriptionSyncInterval != time.Duration(0) {
```
Comment (nathanielc) [inline services/influxdb/service.go]: Here too...

**Comment by nathanielc** [comment]
@titilambert Just a few small fixes but overall looks ready. Once those things are fixed I'll test locally.

**Comment by titilambert** [comment]
@nathanielc done.
I will squash it when you want.

**Comment by titilambert** [comment]
@nathanielc I got a timeout on circle. I can not see where restart the build. Could you do that ?

File: services/influxdb/config.go [hunk 5]
```diff
@@ -34,15 +35,16 @@ type Config struct {
...
+	Timeout                  toml.Duration       `toml:"timeout"`
+	DisableSubscriptions     bool                `toml:"disable-subscriptions"`
+	SubscriptionProtocol     string              `toml:"subscription-protocol"`
+	Subscriptions            map[string][]string `toml:"subscriptions"`
+	ExcludedSubscriptions    map[string][]string `toml:"excluded-subscriptions"`
+	UDPBind                  string              `toml:"udp-bind"`
+	UDPBuffer                int                 `toml:"udp-buffer"`
+	UDPReadBuffer            int                 `toml:"udp-read-buffer"`
+	StartUpTimeout           toml.Duration       `toml:"startup-timeout"`
+	SubscriptionSyncInterval toml.Duration       `toml:"subscriptions-retry-interval"`
```
Comment (nathanielc) [inline services/influxdb/config.go]: The config options still says `retry`.

File: services/influxdb/config.go [hunk 6]
```diff
@@ -73,6 +76,9 @@ func (c *Config) SetDefaultValues() {
 	if c.SubscriptionProtocol == "" {
 		c.SubscriptionProtocol = DefaultSubscriptionProtocol
 	}
+	if c.SubscriptionSyncInterval == toml.Duration(0) {
```
Comment (nathanielc) [inline services/influxdb/config.go]: And this one too.

**Comment by nathanielc** [comment]
@titilambert I think we are close:

1. Can you add a CHANGELOG entry under the 0.14 features section?
2. Can you update the example config in etc/kapacitor/kapacitor.conf with a simple explanation about the new config setting
3. Can you change [this](https://github.com/titilambert/kapacitor/blob/23cc44d6107dcb21a4c5d1c1a04f0d9d00dea41c/services/influxdb/service.go#L250) log to a `D!` debug log since its a much more frequent log.

Testing locally its working great. Thanks!

**Comment by titilambert** [comment]
@nathanielc BTW 1m0s as default value is OK for you ?

**Comment by nathanielc** [comment]
@titilambert Yes, 1m seems like a reasonable default.

**Comment by nathanielc** [comment]
@titilambert Thanks! I'll merge on green.

**Comment by titilambert** [comment]
Thanks ! I will squash it just before ;)

**Comment by nathanielc** [comment]
@titilambert Github can now squash automatically before a merge ;) I got this. Thanks again!

**Comment by titilambert** [comment]
Cool ! Thanks !

**Pull request merged**